routing.go

```diff
@@ -0,0 +1,64 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	cid "github.com/ipfs/go-cid"
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+	routing "github.com/libp2p/go-libp2p-routing"
+	mh "github.com/multiformats/go-multihash"
+)
+
+// RoutingDiscovery is an implementation of discovery using ContentRouting
+type RoutingDiscovery struct {
+	router routing.ContentRouting
+}
+
+func NewRoutingDiscovery(router routing.ContentRouting) Discovery {
+	return &RoutingDiscovery{router}
+}
+
+func (d *RoutingDiscovery) Advertise(ctx context.Context, ns string, opts ...Option) (time.Duration, error) {
+	cid, err := nsToCid(ns)
+	if err != nil {
+		return 0, err
+	}
+
+	err = d.router.Provide(ctx, cid, true)
+	if err != nil {
+		return 0, err
+	}
+
+	// this is the dht provide validity
+	return 24 * time.Hour, nil
+}
+
+func (d *RoutingDiscovery) FindPeers(ctx context.Context, ns string, opts ...Option) (<-chan pstore.PeerInfo, error) {
+	options := &Options{}
```

Comment:
nit: `var options Options`. No need for the indirection (although escape analysis should catch this).

Reply:
ok